tempodb/backend/azure/azure.go

```diff
@@ -0,0 +1,414 @@
+			Prefix:  tenantID + Dir,
+			Details: blob.BlobListingDetails{},
+		})
+		if err != nil {
+			warning = err
+			continue
+		}
+		marker = list.NextMarker
+
+		for _, blob := range list.Segment.BlobPrefixes {
+			idString := strings.TrimSuffix(strings.TrimPrefix(blob.Name, tenantID+"/"), "/")
+			blockID, err := uuid.Parse(idString)
+			if err != nil {
+				warning = fmt.Errorf("failed parse on blockID %s: %v", idString, err)
+				continue
+			}
+			blocks = append(blocks, blockID)
+		}
+		// Continue iterating if we are not done.
+		if !marker.NotDone() {
+			break
+		}
+
+	}
+	return blocks, warning
+}
+
+func (rw *readerWriter) BlockMeta(ctx context.Context, blockID uuid.UUID, tenantID string) (*encoding.BlockMeta, error) {
+
+	name := util.MetaFileName(blockID, tenantID)
+
+	bytes, err := rw.readAll(ctx, name)
+	if err != nil {
+		return nil, backend.ErrMetaDoesNotExist
+	}
+
+	out := &encoding.BlockMeta{}
+	err = json.Unmarshal(bytes, out)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (rw *readerWriter) Bloom(ctx context.Context, blockID uuid.UUID, tenantID string, shardNum int) ([]byte, error) {
+	span, derivedCtx := opentracing.StartSpanFromContext(ctx, "azure.Bloom")
+	defer span.Finish()
+
+	name := util.BloomFileName(blockID, tenantID, shardNum)
+	return rw.readAll(derivedCtx, name)
+}
+
+func (rw *readerWriter) Index(ctx context.Context, blockID uuid.UUID, tenantID string) ([]byte, error) {
+	span, derivedCtx := opentracing.StartSpanFromContext(ctx, "azure.Index")
+	defer span.Finish()
+
+	name := util.IndexFileName(blockID, tenantID)
+	return rw.readAll(derivedCtx, name)
+}
+
+func (rw *readerWriter) Object(ctx context.Context, blockID uuid.UUID, tenantID string, start uint64, buffer []byte) error {
+	span, derivedCtx := opentracing.StartSpanFromContext(ctx, "azure.Object")
+	defer span.Finish()
+
+	name := util.ObjectFileName(blockID, tenantID)
+	err := rw.readRange(derivedCtx, name, int64(start), buffer)
+
+	return err
+}
+
+func (rw *readerWriter) Shutdown() {
+
+}
+
+func (rw *readerWriter) writeAll(ctx context.Context, name string, b []byte) error {
+	err := rw.writer(ctx, bytes.NewReader(b), name)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rw *readerWriter) append(ctx context.Context, src []byte, name string) (string, error) {
+	appendBlobURL := rw.containerURL.NewAppendBlobURL(name)
+
+	resp, err := appendBlobURL.AppendBlock(ctx, bytes.NewReader(src), blob.AppendBlobAccessConditions{}, nil)
+
+	if err != nil {
+		return "", errors.Errorf("cannot upload Azure blob, address: %s", name)
+	}
+	return resp.RequestID(), nil
+
+}
+
+func (rw *readerWriter) writer(ctx context.Context, src io.Reader, name string) error {
+	blobURL := rw.containerURL.NewBlockBlobURL(name)
+
+	if _, err := blob.UploadStreamToBlockBlob(ctx, src, blobURL,
+		blob.UploadStreamToBlockBlobOptions{
+			BufferSize: 3 * 1024 * 1024,
+			MaxBuffers: 4,
+		},
+	); err != nil {
```

Comment:
Return the original error. Wrapping it with a friendly message is ok, but the original error in this block is hidden and would be tough to troubleshoot.